### PR TITLE
Https frontend grouping by vhost

### PR DESCRIPTION
**Pull request by gurinderu**

### CODE REVIEW

**Comment by gurinderu** [comment]
WIP

**Comment by gurinderu** [comment]
Ready

**Comment by drewkerrigan** [comment]
What happened to CI testing?

**Comment by drewkerrigan** [comment]
@mesosphere-ci: retest this please

**Comment by gurinderu** [comment]
@mesosphere-ci: retest this please

**Comment by drewkerrigan** [comment]
@mesosphere-ci: re-test this please

File: marathon_lb.py [hunk 1]
```diff
@@ -427,12 +432,13 @@ def config(apps, groups, bind_http_https, ssl_certs, templater,
...
                                      haproxy_dir,
-                                     duplicate_map)
+                                     duplicate_map,
+                                     https_grouped_frontend_list)
             http_frontend_list.append((backend_weight, p_fe))
             https_frontend_list.append((backend_weight, s_fe))
 
         # if app mode is http, we add the app to the second http frontend
-        # selecting apps by http header X-Marathon-App-Id
+        # selecting apps by http header Xtests/test_marathon_lb.py-Marathon-App-Id
```
Comment (jdef) [inline marathon_lb.py]: ??? accidental change

File: marathon_lb.py [hunk 2]
```diff
@@ -700,27 +732,28 @@ def reloadConfig():
 
 def generateHttpVhostAcl(
         templater, app, backend, haproxy_map, map_array,
-        haproxy_dir, duplicate_map):
+        haproxy_dir, duplicate_map, https_grouped_frontend_list):
```
Comment (jdef) [inline marathon_lb.py]: is this new param an "output" param? it looks like this func mutates its value, which seems surprising. is that an idiom that we want to encourage in this codebase?

File: marathon_lb.py (hunk 1 above)
Comment (gurinderu) [reply marathon_lb.py]: Yep, I will fix that

File: marathon_lb.py (hunk 2 above)
Comment (gurinderu) [reply marathon_lb.py]: It isn't a pure function because this function also mutates a duplicate_map.
But I agree with you and will fix that

**Comment by gurinderu** [comment]
@jdef could you review again? I fixed your comments

**Comment by drewkerrigan** [comment]
@mesosphere-ci: re-test this please

File: Dockerfile [hunk 3]
```diff
@@ -13,7 +13,8 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
         python3 \
         runit \
         gnupg-agent \
-	socat \
+	    socat \
+	    libdpkg-perl \
```
Comment (drewkerrigan) [inline Dockerfile]: This indentation doesn't match, and I'm not sure `libdpkg-perl` is actually needed anywhere?

File: Dockerfile [hunk 4]
```diff
@@ -57,10 +58,10 @@ RUN set -x \
     ' \
     && apt-get update \
     && apt-get install -y --no-install-recommends $buildDeps \
-    && rm -rf /var/lib/apt/lists/* \
-    \
+    && rm -rf /var/lib/apt/lists/*
+
 # Build HAProxy
-    && wget -O haproxy.tar.gz "https://www.haproxy.org/download/$HAPROXY_MAJOR/src/haproxy-$HAPROXY_VERSION.tar.gz" \
+RUN wget -O haproxy.tar.gz "https://www.haproxy.org/download/$HAPROXY_MAJOR/src/haproxy-$HAPROXY_VERSION.tar.gz" \
```
Comment (drewkerrigan) [inline Dockerfile]: Please revert these changes so that it is one large RUN statement - if you would like to refactor the Dockerfile, it can be done in a separate PR.

File: config.py [hunk 5]
```diff
@@ -155,11 +155,62 @@ def load(self):
   mode http
 ''',
                            overridable=False,
-                           description='''\
+                           description='''
```
Comment (drewkerrigan) [inline config.py]: Why was the backslash removed?

File: config.py [hunk 6]
```diff
@@ -845,7 +896,7 @@ def get_descriptions(self):
                 if label.name == 'GROUP':
                     # this one is a special snowflake
                     spec = "`HAPROXY_{n}_" + label.name + "`" + " or " + \
-                        "`HAPROXY_" + label.name + "`"
+                           "`HAPROXY_" + label.name + "`"
```
Comment (drewkerrigan) [inline config.py]: Unnecessary whitespace change

File: config.py [hunk 7]
```diff
@@ -921,9 +988,9 @@ def haproxy_backend_redirect_http_to_https(self, app):
         return self.t['BACKEND_REDIRECT_HTTP_TO_HTTPS'].value
 
     def haproxy_backend_redirect_http_to_https_with_path(self, app):
-        if 'HAPROXY_{0}_BACKEND_REDIRECT_HTTP_TO_HTTPS_WITH_PATH' in\
-          app.labels:
-            return app.\
+        if 'HAPROXY_{0}_BACKEND_REDIRECT_HTTP_TO_HTTPS_WITH_PATH' in \
+                app.labels:
```
Comment (drewkerrigan) [inline config.py]: Please separate whitespace changes from this PR.

File: tests/test_marathon_lb.py [hunk 8]
```diff
@@ -833,7 +833,7 @@ def test_config_simple_app_multiple_vhost_with_path_and_auth(self):
   http-request auth realm "realm" if { ssl_fc_sni test.example.com } \
 path_nginx_10000 !auth_test_example_com_nginx
   use_backend nginx_10000 if { ssl_fc_sni test.example.com } ''' + \
-                                      '''path_nginx_10000
```
Comment (drewkerrigan) [inline tests/test_marathon_lb.py]: Please separate whitespace changes from this PR.
